src/init.cpp

```diff
@@ -793,6 +793,31 @@ void InitParameterInteraction(ArgsManager& args)
             LogInfo("parameter interaction: -onlynet excludes IPv4 and IPv6 -> setting -dnsseed=0\n");
         }
     }
+
+    // We drop UPnP support but kept the arg as hidden for now to display a friendlier error to user who have the
+    // option in their config, and upgrade the setting to -natpmp. TODO: remove for version 30.0.
+    if (args.IsArgSet("-upnp")) {
+        bool overridden = args.SoftSetBoolArg("-natpmp", true);
+        LogWarning("Option '-upnp' is set but UPnP support was dropped in version 29.0.%s",
+                overridden ? " Setting '-natpmp' instead." : "");
+    }
```

Comment:
In commit "init: Handle dropped UPnP support more gracefully" (bd4679a608b2cbc206fa60a180b0fc4ce8a39ec4)

I think it would be better to move the settings.json update code before this code doing "if upnp then softset natpmp". That way soft-setting natpmp only happens if -upnp is enabled in `bitcoin.conf` or on the command line, not if it is enabled in settings.json. So the separate cases will be handled independently and not interacting in a more complicated way.